### ISSUE TITLE
Ineffective mark-compacts near heap limit Allocation failed during multi project build

### ISSUE THREAD

**Issue by devshorts** [microsoft/TypeScript#35534]
**TypeScript Version:**  3.7.2

**Search Terms:** Out Of Memory, Ineffective mark-compacts near heap limit Allocation failed

**Expected behavior:**
I am seeing intermit out of memory errors on a small (private) monorepo:

```
$ npx tsc -b

<--- Last few GCs --->

[16668:0x10280e000]    50509 ms: Scavenge 1389.2 (1422.8) -> 1388.5 (1423.3) MB, 4.2 / 0.0 ms  (average mu = 0.221, current mu = 0.195) allocation failure
[16668:0x10280e000]    50520 ms: Scavenge 1389.5 (1423.3) -> 1388.7 (1424.3) MB, 3.9 / 0.0 ms  (average mu = 0.221, current mu = 0.195) allocation failure
[16668:0x10280e000]    50528 ms: Scavenge 1389.5 (1424.3) -> 1388.7 (1424.8) MB, 3.4 / 0.0 ms  (average mu = 0.221, current mu = 0.195) allocation failure


<--- JS stacktrace --->

==== JS stack trace =========================================

Security context: 0x0191c3c9e6e9 <JSObject>
    0: builtin exit frame: splice(this=0x01917b5bd009 <JSArray[0]>,0x0191dd43b031 <Type map = 0x19188a3b461>,0,0,0x01917b5bd009 <JSArray[0]>)

    1: addTypeToUnion(aka addTypeToUnion) [0x191dd4428a1] [/Users/akropp/src/copper/copper/node_modules/typescript/lib/tsc.js:~35707] [pc=0x3e0aa36a897a](this=0x01916ea026f1 <undefined>,typeSet=0x01917b5bd009 <JSArray[0]>,includes=0,type=0x0191dd43b031 <T...

FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory
 1: 0x10003d035 node::Abort() [/usr/local/bin/node]
 2: 0x10003d23f node::OnFatalError(char const*, char const*) [/usr/local/bin/node]
 3: 0x1001b8e15 v8::internal::V8::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*, bool) [/usr/local/bin/node]
 4: 0x100586d72 v8::internal::Heap::FatalProcessOutOfMemory(char const*) [/usr/local/bin/node]
 5: 0x100589845 v8::internal::Heap::CheckIneffectiveMarkCompact(unsigned long, double) [/usr/local/bin/node]
 6: 0x1005856ef v8::internal::Heap::PerformGarbageCollection(v8::internal::GarbageCollector, v8::GCCallbackFlags) [/usr/local/bin/node]
 7: 0x1005838c4 v8::internal::Heap::CollectGarbage(v8::internal::AllocationSpace, v8::internal::GarbageCollectionReason, v8::GCCallbackFlags) [/usr/local/bin/node]
 8: 0x10059015c v8::internal::Heap::AllocateRawWithLigthRetry(int, v8::internal::AllocationSpace, v8::internal::AllocationAlignment) [/usr/local/bin/node]
 9: 0x1005901df v8::internal::Heap::AllocateRawWithRetryOrFail(int, v8::internal::AllocationSpace, v8::internal::AllocationAlignment) [/usr/local/bin/node]
10: 0x10055f576 v8::internal::Factory::NewFixedArrayWithFiller(v8::internal::Heap::RootListIndex, int, v8::internal::Object*, v8::internal::PretenureFlag) [/usr/local/bin/node]
11: 0x100507088 v8::internal::(anonymous namespace)::ElementsAccessorBase<v8::internal::(anonymous namespace)::FastPackedObjectElementsAccessor, v8::internal::(anonymous namespace)::ElementsKindTraits<(v8::internal::ElementsKind)2> >::Splice(v8::internal::Handle<v8::internal::JSArray>, unsigned int, unsigned int, v8::internal::Arguments*, unsigned int) [/usr/local/bin/node]
12: 0x100245c99 v8::internal::Builtin_Impl_ArraySplice(v8::internal::BuiltinArguments, v8::internal::Isolate*) [/usr/local/bin/node]
13: 0x3e0aa255bf9d
14: 0x3e0aa36a897a
15: 0x3e0aa347cd67
error Command failed with signal "SIGABRT".
```
Re-running the build command and things succeed.  

Running the extended diagnostics gives:

```
$ yarn compile --extendedDiagnostics
yarn run v1.19.1
$ npx tsc -b --extendedDiagnostics
Files:                        541
Lines:                     393492
Nodes:                    1113461
Identifiers:               403759
Symbols:                   287563
Types:                      21804
Memory used:              431266K
Assignability cache size:   10716
Identity cache size:          327
Subtype cache size:          2554
I/O Read time:              0.25s
Parse time:                 1.81s
Program time:               2.54s
Bind time:                  0.80s
Check time:                 1.65s
transformTime time:         0.32s
commentTime time:           0.03s
printTime time:             0.46s
Emit time:                  0.46s
Total time:                 5.44s
Files:                        229
Lines:                     144078
Nodes:                     516607
Identifiers:               188055
Symbols:                   166932
Types:                      15812
Memory used:              628401K
Assignability cache size:    3503
Identity cache size:          175
Subtype cache size:           133
I/O Read time:              0.10s
Parse time:                 0.54s
Program time:               0.95s
Bind time:                  0.34s
Check time:                 0.87s
transformTime time:         0.01s
commentTime time:           0.00s
printTime time:             0.02s
Emit time:                  0.02s
Total time:                 2.19s
✨  Done in 9.27s.
```

However, I have no idea how to provide better debugging information as to why its causing this.

**Comment by RyanCavanaugh** [comment]
We need some way to reproduce this

**Comment by devshorts** [comment]
@RyanCavanaugh i'd love to provide something but its really inconsistent and we can't provide access to our internal company repository.  Is there any debugging or other information I could try and get together to help diagnose maybe a smaller reproduction?

**Comment by ltlian** [comment]
Apologies if this is off topic; We were seeing this on a fairly vanilla create-react-app project. I don't know if the cause is related, but I'll share since the symptom is exactly what we were seeing down to the verbatim error message and we had some trouble finding a fix that was relevant to our project.

Another symptom was that the TS language server would go silent in VSC. No errors, but also no hover definitions or go-to functionality. I didn't check the CPU usage from simply having the project open. It was however 100% until it crashed with the above OOM error when trying to build or start the dev server.

We tracked it down to our fairly large exportless .d.ts file which is generated based on our domain model - a graphQL schema with relay support. One of the types generated in this file shares a name with one from the typescript library ('[Node](https://github.com/microsoft/TypeScript/blob/master/lib/lib.dom.d.ts#L10752)'), so all the types in our file that inherited from 'Node' (quite a lot, as this is the base type for relay paging support) would inherit from the typescript lib.dom type rather than our own. I assume this caused so much overhead that it dies before it manages to report on it.

Once we exported/imported our types normally everything worked as expected.
